src/macaron/policy_engine/souffle.py

```diff
@@ -0,0 +1,187 @@
+# Copyright (c) 2023 - 2023, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""
+Wrapper classes for invoking souffle by subprocess and getting the resulting tables.
+
+Implements a context manager to create and clean up temporary directories.
+"""
+
+import csv
+import glob
+import logging
+import os
+import shutil
+import subprocess  # nosec B404
+import tempfile
+from types import TracebackType
+from typing import Optional
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+class SouffleError(Exception):
+    """Occurs when the souffle program contains errors, or there is an error invoking souffle."""
+
+    # TODO: Use generic Macaron error class
+
+    def __init__(
+        self, command: Optional[list[str] | str] = None, message: str = "An error occurred with calling Souffle."
+    ):
+        self.message = message
+        self.command = command
+        super().__init__(self.message)
+
+
+class SouffleWrapper:
+    """Wrapper class for managing the temporary working directory of the souffle interpreter.
+
+    Examples
+    --------
+    with SouffleWrapper(fact_dir="facts", output_dir="output") as sfl:
+        text = "<souffle program>"
+        result = sfl.interpret_text(text)
+        assert result == {"path": [["1", "2"], ["1", "3"], ["2", "3"]]}
+
+    """
+
+    # The Souffle command
+    _souffle: str
+    # The directory to store outputted facts in
+    output_dir: str
+    # The directory for Souffle to search for datalog files in when using #include
+    include_dir: str
+    # The directory for Souffle to load facts from
+    fact_dir: str
+    # The directory to link Souffle functor shared libraries from
+    library_dir: str
+    souffle_stdout: Optional[str]
+    souffle_stderr: Optional[str]
+
+    # The temporary file to store the datalog program in when executing from the temporary directory
+    TEMP_SOURCEFILE_NAME = "source.dl"
```

Comment:
We should move these to [`defaults.ini`](https://github.com/oracle-samples/macaron/blob/database-policy/src/macaron/config/defaults.ini) and avoid hardcoding them.

Reply:
https://github.com/oracle-samples/macaron/pull/46/commits/db11e761070bbd8c28841a19da111817ef1432ea Am using a temporary file now to avoid having a default name at all.